### DISCUSSION TITLE
slow but steady performance degradation

### DISCUSSION THREAD

**Discussion by lightyear15** [osmandapp/OsmAnd#19304] [General]
Hello there,
I am a long time user of osm for android. I have been using it on an old android phone repurposed as navigation system for my motorbike.
In the years I have been witnessing a slow but steady degradation of performances, especially when it comes to routing calculation, up to the point new, where the app crashes as soon as the destination is further than a few kilometers from my current position.

I must admit that the phone I am using is an old one with very limited resources, RAM in particular, but it is undeniable that a device perfectly capable of running osmAnd a few years ago, has now become obsolete. Nothing has changed in the environment where the app is running, no new app installed, no new services added in the background, so it must be OsmAnd requiring more resources to fulfill the same task.

As a fairly experienced ( C++ ) software engineer, the first thing that came to my mind after this realization was "let's go see the code". I was gladly surprised to understand that the core of Osm is written in C++ with some swig bindings, two technologies I consider myself familiar with.

To the point: I see the C++ codebase has some inefficiencies and code can be improved, both from a maintainability and from runtime performance perspectives. 

Would you like some help in that?

**Comment by Sonwon1** [comment]
Give this guy a seat at the table and see if he can help!

**Comment by vshcherb** [comment]
Hi, it is indeed tricky with C++. As code base grows obviously some degradation going to happen. I think the most important to focus what bothers you the most and where you think it will be interesting for you to contribute. 

Obviously is not an easy process cause somebody will need to accept and maintain that code, so it's better to start with something critical for users and obvious improvements for developers. As I saw in the past large pull requests were stuck cause nobody wanted to take ownership of new functionality but with fixes it goes much easier.

**Reply by lightyear15** [reply]
Thanks @vshcherb 
as you rightly point it out, code degradation is hardly inevitable in growing codebases.

I don't think it would be easy (if not impossible) to spot the line(s) that affect the performance degradation I was talking about, so although I understand your point, I don't think this strategy is feasible.

On the other hand, surfing the codebase I see many points where C++ code can be improved by applying simple code best practices without going through large PRs for overwhelming refactors.

let me give you some examples:
(quick disclaimer: I haven't dug that deep down the code I am about to mention, so pardon me if there are specific design choices that led to such code)

- [PrivateImplementation](https://github.com/osmandapp/OsmAnd-core/blob/master/include/OsmAndCore/PrivateImplementation.h#L23):  ``shared_ptr`` can be replaced with  ``unique_ptr``, being a private implementation there probably is no need to share it. ``unique_ptr`` are much lighter that ``shared_ptr`` both in terms of memory footprint and in runtime effort in creating it. Also safer code as it's impossible to share, i.e. no race condition guaranteed.
- empty destructors: lots of line of code for defining empty destructors, C++11 introduced keyword ``= default`` to define empty constructors/destructors. A fair bit of lines of code can possibily be removed from cpp files. How can maintainers not accept such PRs where loc is going to get reduced?
- constructors input arguments passed by reference even if they get copied over like [here](https://github.com/osmandapp/OsmAnd-core/blob/master/include/OsmAndCore/FavoriteLocationsCollection.h#L28): in such cases, passing by value opens the door to performance gains. In the case where the caller does not need to keep a copy of the inputs, the objects can be ``move``d rather than copied.
- function returning ``shared_ptr``, it's common practice to return a ``unique_ptr`` and let the caller replace it with a ``shared_ptr`` if needed.
- heavy use of Qt containers, standard library containers are, in general, much more efficient and up-to-date with the latest compiler optimization techniques
- using ``new`` rather than ``make_unique`` or ``make_shared``, big source of potential memory leak
- passing ``const float *`` rather than ``std::optional<float>`` to indicate a potential empty/unavailable value, potential source of segmentation fault
...

Also

another fundamental point, prerequisite before embarking in this journey is CI. 
It would be nice to set-up a github workflow that, at least, compiles the library and its java bindings in linux platform, and possible runs the tests. I don't see any.

**Reply by h-h-** [reply]
@lightyear15  If you can afford the time, please do it. Make your PRs! I also see performance issues increasing from update to update. Just starting osmand lasts way too long. It then is very slow e.g. when zooming in/out. Every performance improvement would be greatly appreciated.

**Reply by lightyear15** [reply]
Thanks @h-h-  for the comment

As first step though, I need a huge help for setting up the CI, that is something I consider key to let these changes happen and yet I don't know what the building and testing process is.

Can someone point me to some recipe for building osmAndCore?

**Reply by vshcherb** [reply]
Hello yes it should be straightforward with documentation. Unfortunately there is not many utilities with core so most of it could be tested only with android or ios. Please check android/ios documentation

**Reply by vshcherb** [reply]
I agree with shared ptr and make new cause we had many issues with it already. Some parts of the code are very tricky like in a routing but there are places where we just need to follow certain guidelines

**Comment by lightyear15** [comment]
Please let me know if there is any interest in PRs like these
[716](https://github.com/osmandapp/OsmAnd-core/pull/716)
[743](https://github.com/osmandapp/OsmAnd-core/pull/743)

I tried setting up a CI for osmad-core but the build system is so entangled with other repositories, configuration and variables specific to the build process of external repositories that is very hard to even isolate a small subset of the codebase and have it compiled with no external dependencies. [736](https://github.com/osmandapp/OsmAnd-core/pull/736)

**Reply by vshcherb** [reply]
Yes, I will forward to main core contributors @Andrewschenko
